File: extensions/board/frontend/go/tui/board.go
package board

import (
	"encoding/binary"
	"fmt"
)

// Board is the extension-owned copy of the experimental Board semantic payload. It was moved out of go/tui core so Charm parity does not require Board-specific rendering.
type Board struct {
	Visible       bool
	FocusedCardID uint32
	FilterMode    bool
	FilterText    string
	Cards         []BoardCard
}

// BoardCard is one Board card in the extension-owned legacy GUI payload.
type BoardCard struct {
	ID          uint32
	Status      byte
	Flags       byte
	Task        string
	Model       string
	Timestamp   uint32
	RecentFiles []string
}

// DecodeBoard decodes the legacy gui_board payload owned by the Board extension.
func DecodeBoard(payload []byte) (Board, string, int) {
	if len(payload) < 11 {
		return Board{}, "", len(payload)
	}
	count := int(binary.BigEndian.Uint16(payload[6:8]))
	board := Board{Visible: payload[1] != 0, FocusedCardID: binary.BigEndian.Uint32(payload[2:6]), FilterMode: payload[8] != 0}
	var ok bool
	offset := 9
	board.FilterText, offset, ok = readString16(payload, offset)
	if !ok {
		return board, "", len(payload)
	}
	board.Cards = make([]BoardCard, 0, count)
	for i := 0; i < count; i++ {
		if len(payload) < offset+11 {
			return board, "", len(payload)
		}
		card := BoardCard{ID: binary.BigEndian.Uint32(payload[offset : offset+4]), Status: payload[offset+4], Flags: payload[offset+5]}
		offset += 6
		card.Task, offset, ok = readString16(payload, offset)
		if !ok || len(payload) < offset+5 {
			return board, "", len(payload)
		}
		modelLen := int(payload[offset])
		offset++
		if len(payload) < offset+modelLen+5 {
			return board, "", len(payload)
		}
		card.Model = string(payload[offset : offset+modelLen])
		offset += modelLen
		card.Timestamp = binary.BigEndian.Uint32(payload[offset : offset+4])
		offset += 4
		fileCount := int(payload[offset])
		offset++
		card.RecentFiles = make([]string, 0, fileCount)
		for j := 0; j < fileCount; j++ {
			file, next, ok := readString16(payload, offset)
			if !ok {
				return board, "", len(payload)
			}
			card.RecentFiles = append(card.RecentFiles, file)
			offset = next
		}
		if len(payload) < offset+1 {
			return board, "", len(payload)
		}
		sparklineCount := int(payload[offset])
		offset++
		if len(payload) < offset+sparklineCount*2 {
			return board, "", len(payload)
		}
		offset += sparklineCount * 2
		board.Cards = append(board.Cards, card)
	}
	return board, fmt.Sprintf("%d cards", len(board.Cards)), offset
}

func readString16(payload []byte, offset int) (string, int, bool) {
	if len(payload) < offset+2 {
		return "", offset, false
	}
	length := int(binary.BigEndian.Uint16(payload[offset : offset+2]))
	offset += 2
	if len(payload) < offset+length {
		return "", offset, false
	}
	return string(payload[offset : offset+length]), offset + length, true
}

File: extensions/board/frontend/rust/tui/src/board.rs
//! Extension-owned Board payload support moved out of the Rust TUI core parity path.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Board {
    pub visible: u8,
    pub focused_card_id: u32,
    pub filter_mode: u8,
    pub filter_text: String,
    pub cards: Vec<BoardCard>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BoardCard {
    pub id: u32,
    pub status: u8,
    pub flags: u8,
    pub task: String,
    pub model: String,
    pub timestamp: u32,
    pub recent_files: Vec<String>,
}

pub fn decode_board(bytes: &[u8]) -> Result<(Board, usize), &'static str> {
    if bytes.len() < 11 {
        return Err("short board");
    }

    let visible = bytes[1];
    let focused_card_id = u32::from_be_bytes([bytes[2], bytes[3], bytes[4], bytes[5]]);
    let card_count = u16::from_be_bytes([bytes[6], bytes[7]]) as usize;
    let filter_mode = bytes[8];
    let (filter_text, mut offset) = read_string16(bytes, 9)?;
    let mut cards = Vec::with_capacity(card_count);

    for _ in 0..card_count {
        if bytes.len() < offset + 6 {
            return Err("short board card");
        }
        let id = u32::from_be_bytes([
            bytes[offset],
            bytes[offset + 1],
            bytes[offset + 2],
            bytes[offset + 3],
        ]);
        let status = bytes[offset + 4];
        let flags = bytes[offset + 5];
        offset += 6;

        let (task, next) = read_string16(bytes, offset)?;
        offset = next;
        if bytes.len() < offset + 5 {
            return Err("short board card metadata");
        }
        let model_len = bytes[offset] as usize;
        offset += 1;
        if bytes.len() < offset + model_len + 5 {
            return Err("short board card model");
        }
        let model = String::from_utf8_lossy(&bytes[offset..offset + model_len]).to_string();
        offset += model_len;
        let timestamp = u32::from_be_bytes([
            bytes[offset],
            bytes[offset + 1],
            bytes[offset + 2],
            bytes[offset + 3],
        ]);
        offset += 4;
        let file_count = bytes[offset] as usize;
        offset += 1;
        let mut recent_files = Vec::with_capacity(file_count);
        for _ in 0..file_count {
            let (file, next) = read_string16(bytes, offset)?;
            recent_files.push(file);
            offset = next;
        }
        if bytes.len() < offset + 1 {
            return Err("short board sparkline count");
        }
        let sparkline_count = bytes[offset] as usize;
        offset += 1;
        if bytes.len() < offset + sparkline_count * 2 {
            return Err("short board sparkline");
        }
        offset += sparkline_count * 2;

        cards.push(BoardCard {
            id,
            status,
            flags,
            task,
            model,
            timestamp,
            recent_files,
        });
    }

    Ok((
        Board {
            visible,
            focused_card_id,
            filter_mode,
            filter_text,
            cards,
        },
        offset,
    ))
}

pub fn render_lines(board: &Board, max_height: usize) -> Vec<String> {
    if board.visible == 0 || board.cards.is_empty() || max_height == 0 {
        return Vec::new();
    }

    let mut lines = vec![format!("Board  {} cards", board.cards.len())];
    for card in &board.cards {
        let marker = if card.id == board.focused_card_id || card.flags & 0x02 != 0 {
            ">"
        } else {
            " "
        };
        lines.push(format!(
            "{} {}  {}",
            marker,
            status_name(card.status),
            card.task
        ));
        if lines.len() >= max_height {
            break;
        }
    }
    lines
}

fn read_string16(bytes: &[u8], offset: usize) -> Result<(String, usize), &'static str> {
    if bytes.len() < offset + 2 {
        return Err("short string length");
    }
    let len = u16::from_be_bytes([bytes[offset], bytes[offset + 1]]) as usize;
    let start = offset + 2;
    let end = start + len;
    if bytes.len() < end {
        return Err("short string");
    }
    Ok((String::from_utf8_lossy(&bytes[start..end]).to_string(), end))
}

fn status_name(status: u8) -> &'static str {
    match status {
        0 => "idle",
        1 => "working",
        2 => "iterating",
        3 => "needs you",
        4 => "done",
        5 => "errored",
        _ => "unknown",
    }
}
